### PR TITLE
Improve Hudi connector performance

### LINKED ISSUE #15564: Hudi Connector performance issues while querying partitioned tables
Our team is testing the Hudi Connector on COPY ON WRITE HUDI tables. And we observed that the hudi connector in general performs poorly while querying partitioned tables, with or without metadata enabled. The query supposedly takes time in the state of `Creating InMemory based view for basePath`. Through the coordinator logs we see that query takes 1-2 mins after loading the metadata table to load the file system views using `AbstractTableFileSystemView`.

From seeing the explain plan of the query on Trino we could see that there were no specific filters on the partition column despite providing specific partition column values. Whereas if we do the same through the hive connector then it returns the specified values in the explain plan itself.

**Query:**
```
SELECT id 
FROM hudi.<schema_name>.<table_name>
WHERE insert_ds_ist in ('2022-12-12', '2022-12-11', '2022-12-01', '2022-12-02', '2022-12-03') 
LIMIT 12
```
**Explain Plan:**
`Fragment 0 [SINGLE] Output layout: [id] Output partitioning: SINGLE [] Output[columnNames = [id]] │ Layout: [id:bigint] │ Estimates: {rows: 12 (108B), cpu: 0, memory: 0B, network: 0B} └─ Limit[count = 12] │ Layout: [id:bigint] │ Estimates: {rows: 12 (108B), cpu: 108, memory: 0B, network: 0B} └─ LocalExchange[partitioning = SINGLE] │ Layout: [id:bigint] │ Estimates: {rows: 12 (108B), cpu: 0, memory: 0B, network: 0B} └─ RemoteSource[sourceFragmentIds = [1]] Layout: [id:bigint] Fragment 1 [SOURCE] Output layout: [id] Output partitioning: SINGLE [] LimitPartial[count = 12] │ Layout: [id:bigint] │ Estimates: {rows: 12 (108B), cpu: 108, memory: 0B, network: 0B} └─ TableScan[table = hudi:<schema_name>.<table_name>] Layout: [id:bigint] Estimates: {rows: ? (?), cpu: ?, memory: 0B, network: 0B} id := id:bigint:REGULAR`

**Explain Plain for same query through Hive connector:**
`Fragment 0 [SINGLE] Output layout: [id] Output partitioning: SINGLE [] Output[columnNames = [id]] │ Layout: [id:bigint] │ Estimates: {rows: 12 (108B), cpu: 0, memory: 0B, network: 0B} └─ Limit[count = 12] │ Layout: [id:bigint] │ Estimates: {rows: 12 (108B), cpu: 108, memory: 0B, network: 0B} └─ LocalExchange[partitioning = SINGLE] │ Layout: [id:bigint] │ Estimates: {rows: 12 (108B), cpu: 0, memory: 0B, network: 0B} └─ RemoteSource[sourceFragmentIds = [1]] Layout: [id:bigint] Fragment 1 [SOURCE] Output layout: [id] Output partitioning: SINGLE [] LimitPartial[count = 12] │ Layout: [id:bigint] │ Estimates: {rows: 12 (108B), cpu: 108, memory: 0B, network: 0B} └─ TableScan[table = hive:<schema_name>.<table_name>] Layout: [id:bigint] Estimates: {rows: ? (?), cpu: ?, memory: 0B, network: 0B} id := id:bigint:REGULAR insert_ds_ist:string:PARTITION_KEY :: [[2022-12-01], [2022-12-02], [2022-12-03], [2022-12-11], [2022-12-12]]`

Query runtime:
Hudi Connector: 1 - 2 mins
Hive connector: 4-6 secs

Trino version: 405
HUDI Table version: 0.10.1
Spark Version: 2.4

**Trino Hudi Connector Properties:**
```
connector.name=hudi
hive.metastore.uri={METASTORE_URI}
hive.s3.iam-role={S3_IAM_ROLE}
hive.metastore-refresh-interval=2m
hive.metastore-timeout=3m
hudi.max-outstanding-splits=1000
hive.s3.max-error-retries=50
hive.s3.connect-timeout=1m
hive.s3.socket-timeout=2m
hudi.parquet.use-column-names=true
hudi.metadata-enabled=true
```

**Hudi Properties set while writing:**
```
hoodie.datasource.write.partitionpath.field = "insert_ds_ist",
hoodie.datasource.write.recordkey.field = "id",
hoodie.datasource.write.precombine.field = "_hoodie_incremental_key", (self generated column),
hoodie.datasource.write.hive_style_partitioning = "true",
hoodie.datasource.hive_sync.auto_create_database = "true",
hoodie.parquet.compression.codec = "gzip",
hoodie.table.name = "<table_name>",
hoodie.datasource.write.keygenerator.class = "org.apache.hudi.keygen.SimpleKeyGenerator",
hoodie.datasource.write.table.type = "COPY_ON_WRITE",
hoodie.metadata.enable = "true",
hoodie.datasource.hive_sync.enable = "true",
hoodie.datasource.hive_sync.partition_fields = "insert_ds_ist",
hoodie.datasource.hive_sync.partition_extractor_class = "org.apache.hudi.hive.MultiPartKeysValueExtractor"
```

**General information of table:**
Total rows = 3,784,911,511
Total Partitions = 1700+
Total file objects = ~ 18,000
Total Size on S3 =  ~100 GB

**Coordinator logs for the query:**
[Partitioned_COW_HUDI_Table_Trino400.log](https://github.com/trinodb/trino/files/10323705/Partitioned_COW_HUDI_Table_Trino400.log)

### PR DESCRIPTION
**Pull request by codope**
## Description

Previously, the query execution would wait for all the split generation to complete and splits were loaded in a single thread. With this PR, split generation and processing can happen asynchronously.

## Additional context and related issues

Fixes https://github.com/apache/hudi/issues/7643
Fixes #15564 

## Release notes

(x) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

**Comment by tooptoop4** [comment]
🔥

**Comment by findepi** [comment]
> [martint](https://github.com/martint) requested review from [findepi](https://github.com/findepi) and [hashhar](https://github.com/hashhar) [2 weeks ago](https://github.com/trinodb/trino/pull/16034#event-8482752879)

 @findinpath and @electrum should be the reviewers here.

File: plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/query/HudiReadOptimizedDirectoryLister.java [hunk 1]
```diff
@@ -92,33 +67,29 @@ public List<HudiPartitionInfo> getPartitionsToScan()
                         hiveTable,
                         hiveMetastore))
                 .collect(Collectors.toList());
-
-        return allPartitionInfoList.stream()
-                .filter(partitionInfo -> partitionInfo.getHivePartitionKeys().isEmpty() || partitionInfo.doesMatchPredicates())
```
Comment (ksoullpwk) [inline plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/query/HudiReadOptimizedDirectoryLister.java:97]: Do we still run `partitionInfo.doesMatchPredicates()` someplace? This might help reducing the external loading of `partitionInfo.getHivePartitionKeys()` which calls directly to Hive.

File: plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/partition/HiveHudiPartitionInfo.java [hunk 2]
```diff
@@ -38,8 +37,6 @@
 public class HiveHudiPartitionInfo
         implements HudiPartitionInfo
 {
-    private static final Logger log = Logger.get(HiveHudiPartitionInfo.class);
```
Comment (findinpath) [inline plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/partition/HiveHudiPartitionInfo.java]: Make the cleanup in a separate commit.

**Comment by findinpath** [comment]
There are a lot of changes in this PR and (besides `TestHudiConfig`) no tests to accompanying them. Can you consider adding more test coverage for your changes?

**Comment by codope** [comment]
@findinpath @ksoullpwk Thanks for reviewing the PR. I will address your comments this week.

File: plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/query/HudiReadOptimizedDirectoryLister.java (hunk 1 above)
Comment (codope) [reply plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/query/HudiReadOptimizedDirectoryLister.java:97]: Good catch! We should call it in `HudiPartitionInfoLoader`.

File: plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/partition/HiveHudiPartitionInfo.java (hunk 2 above)
Comment (codope) [reply plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/partition/HiveHudiPartitionInfo.java]: This was already removed by some other commit.

**Comment by codope** [comment]
> There are a lot of changes in this PR and (besides `TestHudiConfig`) no tests to accompanying them. Can you consider adding more test coverage for your changes?

@findinpath Can you please review again? 
Would appreciate if you can help with what kind of tests we can add here. There is no new functionality. This PR is a performance enhancement. Sure, there is some change in the behavior but existing tests will go through that path. So functional correctness is already tested.

**Comment by tooptoop4** [comment]
🦗

**Review by tooptoop4 — APPROVED** [review]

**Comment by codope** [comment]
@findinpath @electrum gentle ping to review the PR.

File: plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiConfig.java [hunk 3]
```diff
@@ -90,36 +91,6 @@ public boolean getUseParquetColumnNames()
         return this.shouldUseParquetColumnNames;
     }
 
-    @Config("hudi.min-partition-batch-size")
-    @ConfigDescription("Minimum number of partitions returned in a single batch.")
```
Comment (findinpath) [inline plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiConfig.java:94]: Generally the unused configurations are marked as defunct to avoid breaking Trino while doing an upgrade

https://github.com/trinodb/trino/blob/dc7a9fd9605cb56d69e16c461a8e8d1ab0645a85/plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveConfig.java#L49-L62

Comment (electrum) [reply plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiConfig.java:94]: Minor clarification: removal of configs still breaks Trino if the user is using them (we fail on unused/unknown configs), but the user gets an explicit error message saying the config is defunct (rather than just unknown).

File: plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/ForHudiSplitSource.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+import static java.lang.annotation.ElementType.FIELD;
+import static java.lang.annotation.ElementType.METHOD;
+import static java.lang.annotation.ElementType.PARAMETER;
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+@Retention(RUNTIME)
+@Target({FIELD, PARAMETER, METHOD})
+@Qualifier
+public @interface ForHudiSplitSource
```
Comment (electrum) [inline plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/ForHudiSplitSource.java:30]: Nit: format empty interface like
```java
public @interface ForHudiSplitSource {}

File: plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiConfig.java [hunk 5]
```diff
@@ -191,4 +162,47 @@ public HudiConfig setMaxOutstandingSplits(int maxOutstandingSplits)
...
+
+    @Min(1)
+    public int getSplitLoaderParallelism()
+    {
+        return splitLoaderParallelism;
+    }
+
+    @Config("hudi.split-loader-parallelism")
+    @ConfigDescription("Number of threads to run background split loader. "
+            + "A single background split loader is needed per query.")
```
Comment (electrum) [inline plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiConfig.java:188]: No need to wrap here

File: plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiConfig.java [hunk 6]
```diff
@@ -191,4 +162,47 @@ public HudiConfig setMaxOutstandingSplits(int maxOutstandingSplits)
...
+        return this;
+    }
+
+    @Min(1)
+    public long getPerTransactionMetastoreCacheMaximumSize()
+    {
+        return perTransactionMetastoreCacheMaximumSize;
+    }
+
+    @LegacyConfig("hive.per-transaction-metastore-cache-maximum-size")
```
Comment (electrum) [inline plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiConfig.java:201]: Did Hudi previously use the Hive config? I don't see that in this PR

File: plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiModule.java [hunk 7]
```diff
@@ -76,7 +80,27 @@ public void configure(Binder binder)
     @Provides
     public ExecutorService createExecutorService()
     {
-        return newCachedThreadPool(daemonThreadsNamed("hudi-split-manager-%d"));
+        return newCachedThreadPool(daemonThreadsNamed("hudi-split-manager-%s"));
+    }
+
+    @ForHudiSplitSource
```
Comment (electrum) [inline plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiModule.java:86]: Put these in the order
```java
@Provides
@Singleton
@ForHudiSplitSource
```
Then it reads naturally as "provides singleton for Hudi split source"

File: plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiModule.java [hunk 8]
```diff
@@ -76,7 +80,27 @@ public void configure(Binder binder)
...
+    @Singleton
+    @Provides
+    public ScheduledExecutorService createSplitLoaderExecutor(HudiConfig hudiConfig)
```
Comment (electrum) [inline plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiModule.java:89]: Why do we need a `ScheduledExecutorService`? It seems to only be used as a regular `ExecutorService`.

Also, why do we need two separate thread pools, both configured with the same concurrency?

File: plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/query/HudiReadOptimizedDirectoryLister.java [hunk 9]
```diff
@@ -62,63 +54,43 @@ public HudiReadOptimizedDirectoryLister(
...
+                                tableHandle.getPartitionPredicates(),
+                                hiveTable,
+                                hiveMetastore)));
     }
 
     @Override
     public List<FileStatus> listStatus(HudiPartitionInfo partitionInfo)
     {
+        if (latestInstant != null) {
+            return fileSystemView.getLatestFileSlicesBeforeOrOn(partitionInfo.getRelativePartitionPath(), latestInstant, false)
```
Comment (electrum) [inline plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/query/HudiReadOptimizedDirectoryLister.java:79]: This seems like new functionality, unrelated to the concurrency change.

Since we're expanding the usage of `FileSystemViewManager`, could you add another commit to this PR as a precursor that switches `HudiReadOptimizedDirectoryLister` to use `TrinoFileSystem`?

Comment (codope) [reply plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/query/HudiReadOptimizedDirectoryLister.java:79]: `TrinoFileSystem` can be used list a partition but a partition in Hudi could contain older file versions as well and to filter those out we need `FileSystemViewManager`. It is more efficient to first filter latest version (based on Hudi commit metadata in [timeline](https://hudi.apache.org/docs/timeline)) and then list the remaining files. 
That said, we can still remove the usage of `FileStatus` from `HudiDrirectoryLister`. I'll create a separate PR for that.

File: plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiConfig.java (hunk 6 above)
Comment (codope) [reply plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiConfig.java:201]: I just reused Hive config. I should change it to `hudi.*`

File: plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiConfig.java (hunk 3 above)
Comment (codope) [reply plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiConfig.java:94]: noted.

File: plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiModule.java (hunk 8 above)
Comment (codope) [reply plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiModule.java:89]: Added two separate thread pools due to the nature of work being done. One task is simply adding partitions to the async queue, the other task is listing those partitions which is IO bound. I kept the default concurrency same for both as it should not matter much for small table (few tens or hundreds of partitions) but for larger tables they should be tuned. It's a way to balance the tradeoff. When working with larger tables with thousands of partitions, it's better to increase the `splitGeneratorParallelism`. 
You're right about `ScheduledExecutorService`. I think we started with that but don't see a need.

**Comment by vinothchandar** [comment]
cc @umehrot2  from Athena and @bvaradar from Robinhood, who were asking about this.

File: plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/partition/HudiPartitionInfoLoader.java [hunk 10]
```diff
@@ -13,110 +13,77 @@
...
+        log.debug("HudiPartitionInfoLoader %s finishes in %d ms", this, timer.endTimer());
     }
 
-    public Deque<HudiPartitionInfo> getPartitionQueue()
+    private void generateSplitsFromPartition(String partitionName)
     {
-        return partitionQueue;
+        Optional<HudiPartitionInfo> partitionInfo = hudiDirectoryLister.getPartitionInfo(partitionName);
+        partitionInfo.ifPresent(hudiPartitionInfo -> {
+            if (hudiPartitionInfo.getHivePartitionKeys().isEmpty() || hudiPartitionInfo.doesMatchPredicates()) {
```
Comment (ksoullpwk) [inline plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/partition/HudiPartitionInfoLoader.java:74]: Can we run `doesMatchPredicates()` before `getHivePartitionKeys().isEmpty()`?
I found this `getHivePartitionKeys()` needs to access to Hive and it takes some network time and makes Hive have more loading. On the other hands, `doesMatchPredicates()` is running to verify the partition internally in CPU (which is faster than the network time) and there is no need to access to Hive anymore if the partition doesn't match the predicates.

**Comment by vinothchandar** [comment]
@ksoullpwk Could you please review again or one of the trino maintainers chime in? We need to get this in, so Hudi users can migrate to the hudi connector away from hive connector.

**Review by ksoullpwk — COMMENTED** [review]
As I see, some configs have changed to this PR. Don't forget to update the document `https://github.com/trinodb/trino/blob/master/docs/src/main/sphinx/connector/hudi.rst`.

**Comment by codope** [comment]
@electrum @findinpath Can you please review again? Added a tested for `HudiPartitionManager` and addressed other comments.

**Comment by electrum** [comment]
Apologies for the delay on this. Can you rebase? We completed the removal of Hadoop from the Hudi connector and can now start merging these other changes.

**Comment by codope** [comment]
@electrum gentle ping

File: plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/partition/HudiPartitionInfoLoader.java (hunk 10 above)
Comment (ksoullpwk) [reply plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/partition/HudiPartitionInfoLoader.java:67]: I'm trying to run again from your branch and putting the log into `ThriftHiveMetastore.getPartition` to check if we have any calls into Hive. It seems like Hive has been called for many time (as the number of partitions).
image

My example is I have a table `xxx.yyy` with column "date" as partition which have more than 500 days of data. When I run `SELECT * FROM hudi.xxx.yyy WHERE date >= '2023-06-01' LIMIT 10`. It tries to call `ThriftHiveMetastore.getPartition` for every partition, but, what we expected, it should call for only around 6-7 partitions.
image

Another question, in the case that `getHivePartitionKeys()` returns empty, this line will become true and then generate split for empty partition key(?), is this what we expected?

Comment (codope) [reply plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/partition/HudiPartitionInfoLoader.java:67]: Sorry, I misunderstood your point earlier. Let me take a look and fix this today.

File: plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/query/HudiReadOptimizedDirectoryLister.java (hunk 9 above)
Comment (codope) [reply plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/query/HudiReadOptimizedDirectoryLister.java]: I've rebased with latest master and used `HudiTableFileSystemView` instead.

File: plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/partition/HudiPartitionInfoLoader.java (hunk 10 above)
Comment (codope) [reply plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/partition/HudiPartitionInfoLoader.java]: @ksoullpwk I've fixed the issue now. Thanks for pointing it out.

**Comment by codope** [comment]
@electrum I've rebased and addressed all comments. Please take a pass again.

**Comment by codope** [comment]
@trinodb/maintainers Please review this PR.

File: plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/partition/HudiPartitionInfoLoader.java (hunk 10 above)
Comment (ksoullpwk) [reply plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/partition/HudiPartitionInfoLoader.java]: Hi @codope, I have tested again after swapping the condition and it seems like the issue still happens. It might be something wrong in this line.
From my understanding, we want to generate the split only if the predicate is in our filter and the partition does really exists. I guess this line should be `hudiPartitionInfo.doesMatchPredicates() && !hudiPartitionInfo.getHivePartitionKeys().isEmpty()`.
So, we check first if it does match predicates, in case of yes, we will call to Hive and check if the partition key does not empty.

File: plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiConfig.java [hunk 11]
```diff
@@ -191,4 +166,45 @@ public HudiConfig setMaxOutstandingSplits(int maxOutstandingSplits)
...
+
+    @Min(1)
+    public int getSplitGeneratorParallelism()
+    {
+        return splitGeneratorParallelism;
+    }
+
+    @Config("hudi.split-generator-parallelism")
+    @ConfigDescription("Number of threads to generate splits from partitions.")
+    public HudiConfig setSplitGeneratorParallelism(int splitGeneratorParallelism)
```
Comment (ksoullpwk) [inline plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiConfig.java:178]: Don't forget to update the document for these new configurations. And also clean the document for obsoleted configs.

File: plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/partition/HudiPartitionInfoLoader.java (hunk 10 above)
Comment (codope) [reply plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/partition/HudiPartitionInfoLoader.java]: @ksoullpwk After flipping the conditional expressions, metastore call should only happen for the partitions that match predicate. The second part OR condition i.e. `hudiPartitionInfo.getHivePartitionKeys().isEmpty()` is to generate splits for non-partitioned table. It was the same way before:
https://github.com/trinodb/trino/blob/5295f3d5a66020f735e993911f743a5df860269f/plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/query/HudiReadOptimizedDirectoryLister.java#L90-L92

You can also check `TestHudiSmokeTest#testReadNonPartitionedTable`

File: plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiConfig.java (hunk 11 above)
Comment (codope) [reply plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiConfig.java:178]: Ack, will create a separate PR.

File: plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/partition/HudiPartitionInfoLoader.java (hunk 10 above)
Comment (codope) [reply plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/partition/HudiPartitionInfoLoader.java]: @ksoullpwk Good catch! I have fixed the check just now. Please take a look.

File: plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiPartitionManager.java [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+    @Inject
+    public HudiPartitionManager(TypeManager typeManager)
+    {
+        this.typeManager = requireNonNull(typeManager, "typeManager is null");
+    }
+
+    public List<String> getEffectivePartitions(HudiTableHandle tableHandle, HiveMetastore metastore)
+    {
+        Optional<Table> table = metastore.getTable(tableHandle.getSchemaName(), tableHandle.getTableName());
+        verify(table.isPresent());
```
Comment (electrum) [inline plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiPartitionManager.java:47]:
```java
Table table = metastore.getTable(...)
        .orElseThrow(() -> new TableNotFoundException(tableHandle.getSchemaTableName()));

File: plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiSessionProperties.java [hunk 13]
```diff
@@ -116,6 +107,21 @@ public HudiSessionProperties(HudiConfig hudiConfig, ParquetReaderConfig parquetR
...
                         },
+                        false),
+                integerProperty(
+                        MAX_SPLITS_PER_SECOND,
+                        "Rate at which splits are enqueued for processing. The queue will throttle if this rate limit is breached.",
+                        hudiConfig.getMaxSplitsPerSecond(),
+                        false),
+                integerProperty(
+                        MAX_OUTSTANDING_SPLITS,
+                        "Maximum outstanding splits in a batch enqueued for processing.",
```
Comment (electrum) [inline plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiSessionProperties.java:118]: Don't end descriptions with a period if they only have a single sentence

**Review by electrum — COMMENTED** [review]
A few comments, otherwise looks good

**Review by electrum — APPROVED** [review]

**Review by ksoullpwk — APPROVED** [review]

File: plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/partition/HudiPartitionInfoLoader.java (hunk 10 above)
Comment (codope) [reply plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/partition/HudiPartitionInfoLoader.java]: Thanks for your help.

File: plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiModule.java [hunk 14]
```diff
@@ -64,22 +67,43 @@ public void configure(Binder binder)
...
+    {
+        return newScheduledThreadPool(
+                hudiConfig.getSplitLoaderParallelism(),
+                daemonThreadsNamed("hudi-split-loader-%s"));
     }
 
+    @Provides
     @Singleton
+    @ForHudiBackgroundSplitLoader
+    public ExecutorService createSplitGeneratorExecutor(HudiConfig hudiConfig)
```
[the remark this answers is not in the document]
Comment (codope) [reply plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiModule.java]: That's right. I've refactored as suggested.

File: plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiPartitionManager.java (hunk 12 above)
Comment (codope) [reply plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiPartitionManager.java]: Fixed in latest commit.

**Comment by codope** [comment]
Thanks @electrum @ksoullpwk for quick feedback. I've updated the docs and addressed other comments. PR is ready to merge.

**Comment by yx-keith** [comment]
this pr use hive metastore's getPartition function when load parition info, it will take a long time if there are many partitions.

**Comment by yx-keith** [comment]
I optimized it in 406 version,  brings 40x to 50x performance improment in my environment

https://github.com/trinodb/trino/commit/743c41414de045be4a9420332cddbfac032f25df

**Comment by codope** [comment]
> I optimized it in 406 version, brings 40x to 50x performance improment in my environment
> 
> [743c414](https://github.com/trinodb/trino/commit/743c41414de045be4a9420332cddbfac032f25df)

@yx-keith This is great! Do you want to raise a PR so that we can port over your changes to the latest Trino version? I can help with the review. If you have some benchmarks on a large dataset (possibly thousands of partitions), feel free to share it as well. Once you have the PR, I can also help with benchmarking against TPC-DS dataset which has about ~2k partitions (good enough to bring out the perf diff).

**Comment by yx-keith** [comment]
yes, thank you,  Is it necessary to merge my code into master branch and then raise a PR? 
I use table and dataset in this issue:
https://github.com/apache/hudi/issues/7643

**Comment by yx-keith** [comment]
> > I optimized it in 406 version, brings 40x to 50x performance improment in my environment
> > [743c414](https://github.com/trinodb/trino/commit/743c41414de045be4a9420332cddbfac032f25df)
> 
> @yx-keith This is great! Do you want to raise a PR so that we can port over your changes to the latest Trino version? I can help with the review. If you have some benchmarks on a large dataset (possibly thousands of partitions), feel free to share it as well. Once you have the PR, I can also help with benchmarking against TPC-DS dataset which has about ~2k partitions (good enough to bring out the perf diff).

yes, thank you, Is it necessary to merge my code into master branch and then raise a PR?
I use table and dataset in this issue:
https://github.com/apache/hudi/issues/7643
there are 744 partitions

**Comment by codope** [comment]
@yx-keith I would suggest to rebase you changes on top of latest master code and run your test with the dataset (744 partitions) that you have again. Ideally, metastore rpc happens for only matching partitions. It was fixed in https://github.com/trinodb/trino/blob/3661df860953167fe1954f0f29f16662f346f4b6/plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/partition/HudiPartitionInfoLoader.java#L67-L68

Also, check this comment where we discussed the additional metastore calls - https://github.com/trinodb/trino/pull/16034#discussion_r1189732900

**Comment by yx-keith** [comment]
> Reference

Ok，
this is my pr:
async and concurrently to process hudi table partitions in batches and create splits #18511

**Comment by yx-keith** [comment]
> @yx-keith I would suggest to rebase you changes on top of latest master code and run your test with the dataset (744 partitions) that you have again. Ideally, metastore rpc happens for only matching partitions. It was fixed in
> 
> https://github.com/trinodb/trino/blob/3661df860953167fe1954f0f29f16662f346f4b6/plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/partition/HudiPartitionInfoLoader.java#L67-L68
> 
> Also, check this comment where we discussed the additional metastore calls - [#16034 (comment)](https://github.com/trinodb/trino/pull/16034#discussion_r1189732900)

ok,  this is my pr
**https://github.com/trinodb/trino/pull/18511**

**Comment by yx-keith** [comment]
> > [martint](https://github.com/martint) requested review from [findepi](https://github.com/findepi) and [hashhar](https://github.com/hashhar) [2 weeks ago](https://github.com/trinodb/trino/pull/16034#event-8482752879)
> 
> @findinpath and @electrum should be the reviewers here.
